integration/lifecycle/capabilities_endorse_test.go

```diff
@@ -0,0 +1,261 @@
+			Sequence:            "1",
+			InitRequired:        true,
+			Label:               "my_simple_chaincode",
+		}
+
+		By("setting up the channel with v3_0 capabilities and without the ed25519 peer")
+		network.CreateAndJoinChannels(orderer)
+		network.UpdateChannelAnchors(orderer, "testchannel")
+		network.VerifyMembership(network.PeersWithChannel("testchannel"), "testchannel")
+		nwo.EnableCapabilities(network, "testchannel", "Application", "V2_0", orderer, network.Peer("Org1", "peer0"), network.Peer("Org2", "peer0"))
+		nwo.EnableChannelCapabilities(network, "testchannel", "V3_0", false, orderer, []*nwo.Orderer{orderer},
+			network.Peer("Org1", "peer0"),
+			network.Peer("Org2", "peer0"),
+		)
+
+		By("joining the ed25519 peer to the channel")
+		ed25519Peer := network.Peer("Org1", "peer1")
+		ed25519Peer.Channels = []*nwo.PeerChannel{
+			{Name: "testchannel", Anchor: false},
+		}
+		network.JoinChannel("testchannel", orderer, ed25519Peer)
+		testPeers = append(testPeers, ed25519Peer)
+
+		By("deploying the chaincode")
+		nwo.PackageChaincodeBinary(chaincode)
+		chaincode.SetPackageIDFromPackageFile()
+
+		nwo.InstallChaincode(network, chaincode, testPeers...)
+
+		By("verifying the installed chaincode package matches the one that was submitted")
+		sess, err := network.PeerAdminSession(testPeers[0], commands.ChaincodeGetInstalledPackage{
+			PackageID:       chaincode.PackageID,
+			OutputDirectory: testDir,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Eventually(sess, network.EventuallyTimeout).Should(gexec.Exit(0))
+		fileBytes, err := ioutil.ReadFile(chaincode.PackageFile)
+		Expect(err).NotTo(HaveOccurred())
+		fileBytesFromPeer, err := ioutil.ReadFile(filepath.Join(network.RootDir, chaincode.PackageID+".tar.gz"))
+		Expect(err).NotTo(HaveOccurred())
+		Expect(fileBytesFromPeer).To(Equal(fileBytes))
+
+		nwo.ApproveChaincodeForMyOrg(network, "testchannel", orderer, chaincode, testPeers...)
+
+		nwo.CheckCommitReadinessUntilReady(network, "testchannel", chaincode, network.PeerOrgs(), testPeers...)
+		nwo.CommitChaincode(network, "testchannel", orderer, chaincode, testPeers[0], testPeers...)
+		nwo.InitChaincode(network, "testchannel", orderer, chaincode, testPeers...)
+
+		By("invoking the chaincode with the ed25519 endorser")
+		endorsers := []*nwo.Peer{
+			network.Peer("Org1", "peer1"),
+			network.Peer("Org2", "peer0"),
+		}
+		RunQueryInvokeQuery(network, orderer, "My_1st-Chaincode", 100, endorsers...)
+
+		By("downgrading the channel capabilities back to v2_0")
+		nwo.EnableChannelCapabilities(network, "testchannel", "V2_0", false, orderer, []*nwo.Orderer{orderer},
+			network.Peer("Org1", "peer0"),
+			network.Peer("Org2", "peer0"),
+		)
+
+		By("invoking the chaincode again, but expecting a failure")
+		RunInvokeAndExpectFailure(network, orderer, "My_1st-Chaincode", "(ENDORSEMENT_POLICY_FAILURE)", endorsers...)
+	})
+})
+
+func changePeerOrOrdererToEd25519(network *nwo.Network, orgName, entityname string) {
```

Comment:
why do you need to read the entire config? Can't you just shut down the node, and replace its MSP folder and start it again?

Reply:
The crypto material on integration tests are created by cryptogen, from the template here: [nwo crypto template](https://github.com/hyperledger/fabric/blob/5cb00bef7c4c499481b5538034a1975dbe3c9319/integration/nwo/template/crypto_template.go#L9-L11). To avoid major changes on the default network definitions, I use the default [MultiNodeEtcdRaft](https://github.com/hyperledger/fabric/blob/5cb00bef7c4c499481b5538034a1975dbe3c9319/integration/nwo/standard_networks.go#L234), but add another peer to have a ed25519 key. Also, I provide an ed25519 key for **orderer3**. Both are enabled by modifying the cryptogen config for the extra peer and **orderer3**, which is the action performed by this function. 

This prevents me from creating another cryptogen template for an ed25519 integration test, or adding another network definition.

Do you think there is a better way?

Reply:
By the way, I am fixing this test. I designed it when there was a system channel.

Reply:
If all you want is to add a peer with a different key and certificate, you can generate the key and certificate and then have the CA sign the key beforehand: https://github.com/hyperledger/fabric/blob/main/integration/raft/cft_test.go#L1167-L1200